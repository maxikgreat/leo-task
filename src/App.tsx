import React, { VFC, useState, useEffect } from 'react';
import axios from 'axios';

import { User, Search } from './components';

export interface UserType {
  id: number,
  name: string,
  username: string
}

interface AppState {
  loading: boolean,
  error: string | null,
  data: UserType[]
}

const App: VFC = () => {
  const [users, setUsers] = useState<AppState>({
    loading: false,
    error: null,
    data: []
  });

  const [search, setSearch] = useState('');

  useEffect(() => {
    setUsers(prevState => ({ ...prevState, loading: true }));
    axios.get('https://jsonplaceholder.typicode.com/users')
      .then(({ data }) => {
        setUsers(prevState => ({ ...prevState, data }));
      })
      .catch((err) => setUsers(prevState => ({ ...prevState, error: err.message })))
      .finally(() => setUsers(prevState => ({ ...prevState, loading: false })));
  }, []);

  const getFilteredUsers = (): UserType[] =>
    users.data.filter(({ name }) =>
      name.toLowerCase().includes(search.toLowerCase())
    );

  return (
    <main>
      <h1>Users list</h1>
      <Search value={search} setValue={setSearch} />
      {users.loading && !users.error && <h2>Loading...</h2>}
      {!users.loading && !users.error && (
        <ol>
          {getFilteredUsers().map((user) => (
            <User key={user.id} user={user} />
          ))}
        </ol>
      )}
      {!users.loading && !users.error && !getFilteredUsers().length && (
        <h2>No users found!</h2>
      )}
      {!users.loading && users.error && (
        <h2 className="error" aria-label="error">{users.error}</h2>
      )}
    </main>
  );
};

export default App;
