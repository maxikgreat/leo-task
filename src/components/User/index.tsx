import React, { VFC } from 'react';
import { UserType } from '../../App';

interface UserProps {
  user: UserType
}

const User: VFC<UserProps> = ({ user }) => {
  return (
    <li>
      <strong>{user.name}</strong> @{user.username}
    </li>
  );
};

export default User;
