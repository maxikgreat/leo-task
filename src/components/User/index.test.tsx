import { render } from '@testing-library/react';

import User from '.';
import { UserType } from '../../App';

describe('User component', () => {
  const user: UserType = {
    id: 1,
    name: 'Maksym Vasylenko',
    username: 'maxikgreat'
  };

  it('should render component correctly and match snapshot', () => {
    const elem = render(<User user={user}/>);

    expect(elem.getByText(user.name)).toBeTruthy();

    expect(elem.getByRole('listitem')).toMatchSnapshot();
  });
});
