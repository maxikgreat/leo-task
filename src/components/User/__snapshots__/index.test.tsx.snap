// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component should render component correctly and match snapshot 1`] = `
<li>
  <strong>
    Maksym Vasylenko
  </strong>
   @
  maxikgreat
</li>
`;
